src/uct/ib/ud/base/ud_iface.h

```diff
@@ -427,8 +426,11 @@ uct_ud_iface_check_grh(uct_ud_iface_t *iface, void *packet, int is_grh_present,
      *     - 0x0 : GRH - (RoCE v1.0)
      *     - 0x1 : IPv6 - (RoCE v1.5/v2.0)
      *     - 0x2 : IPv4 - (RoCE v1.5/v2.0)
+     * Some RoCE drivers fail to set the sl correctly so instead of relying
+     * on the sl the check uses the current address family to detect the
+     * correct gid length.
      */
-    gid_len = ((roce_pkt_type & UCT_IB_CQE_SL_PKTYPE_MASK) == 0x2) ?
+    gid_len = (iface->super.gid_info.roce_info.addr_family == AF_INET) ?
```

Comment:
@evgeny-leksikov, can we take `iface->gid_table.last_len` instead?

Reply:
Had the same question. Figured there was a reason that the code cared about what the address family was instead of using the cached size. I can change it to use that and see if it works on these systems. The address family can not change so the grh should always match and it it doesn't this will silently drop the packet. We were seeing this since most of the packets were being treated as IPv6 even though it is using IPv4. Only 1/7 were being handled correctly.

Reply:
This change actually breaks https://github.com/openucx/ucx/pull/5237 logic: it assumes the length of the received DGID is same as the length of the default GID index on the interface (say, GID[0]). Therefore the uct_ud_gid_equal() result might be incorrect when a non-default GID index is used (with different address family/length)

@hjelmn can we use a different heuristic to deduce the packet GID length, according to the GID contents for example? and do this only for cases where the RoCE driver fail to set the sl correctly?

Reply:
From what I can tell setting the sl in this way is proprietary and only done by Mellanox. We determined this after some more research into the the service level. So, technically it is not incorrect to not set the sl (though I agree that Intel should be initializing it to something instead of not setting it at all). Let me know if that determination is incorrect.

We can probably use the grh itself to deduce the address type. For IPv4 the first 20 bytes are undefined but with IPv6 all 40 bytes are used. Let me look at the structure and see if there is something in the second 20 bytes we can hook into.